_webi/transform-releases.js

```diff
@@ -64,59 +64,79 @@ function createFormatsSorter(formats) {
 async function getCachedReleases(pkg) {
   // returns { download: '<template string>', releases: [{ version, date, os, arch, lts, channel, download}] }
 
-  function putCache() {
-    cache[pkg].promise = cache[pkg].promise.then(function () {
-      var age = Date.now() - cache[pkg].updatedAt;
-      if (age < staleAge) {
-        //console.debug('NOT STALE ANYMORE - updated in previous promise');
-        return cache[pkg].all;
-      }
-      //console.debug('DOWNLOADING NEW "%s" releases', pkg);
-      var pkgdir = path.join(installerDir, pkg);
-      return Releases.get(pkgdir)
-        .then(function (all) {
-          //console.debug('DOWNLOADED NEW "%s" releases', pkg);
-          cache[pkg].updatedAt = Date.now();
-          cache[pkg].all = all;
-        })
-        .catch(function (e) {
-          console.error(
-            'Error fetching releases for "%s": %s',
-            pkg,
-            e.toString(),
-          );
-          cache[pkg].all = { download: '', releases: [] };
-        })
-        .then(function () {
-          return cache[pkg].all;
-        });
-    });
+  async function chainCachePromise(fn) {
+    cache[pkg].promise = cache[pkg].promise.then(fn);
     return cache[pkg].promise;
   }
 
-  var p;
+  async function sleep(ms) {
+    return await new Promise(function (resolve, reject) {
+      setTimeout(resolve, ms);
+    });
+  }
+
+  async function putCache() {
+    var age = Date.now() - cache[pkg].updatedAt;
+    if (age < staleAge) {
+      //console.debug('NOT STALE ANYMORE - updated in previous promise');
+      return cache[pkg].all;
+    }
+
+    //console.debug('DOWNLOADING NEW "%s" releases', pkg);
+    var pkgdir = path.join(installerDir, pkg);
+
+    // workaround for request timeout seeming to not work
```

Comment:
This is the most important bit of the fix - a race between an update and returning the existing data.

This should prevent any promise deadlocks, and any failed request will just hang out in the background until it gets garbage collected.